gr-digital/include/gnuradio/digital/crc.h

```diff
@@ -0,0 +1,86 @@
+/* -*- c++ -*- */
+/*
+ * Copyright 2022 Daniel Estevez <daniel@destevez.net>
+ *
+ * This file is part of GNU Radio
+ *
+ * SPDX-License-Identifier: GPL-3.0-or-later
+ *
+ */
+
+#ifndef INCLUDED_DIGITAL_CRC_H
+#define INCLUDED_DIGITAL_CRC_H
+
+#include <gnuradio/digital/api.h>
+
+#include <stdint.h>
+#include <array>
+#include <vector>
+
+namespace gr {
+namespace digital {
+
+/*!
+ * \brief Calculates a CRC
+ * \ingroup packet_operators_blk
+ *
+ * \details
+ * This class calculates a CRC with configurable parameters.
+ * A table-driven byte-by-byte approach is used in the CRC
+ * computation.
+ */
+class DIGITAL_API crc
+{
+public:
+    /*!
+     * \brief Construct a CRC calculator instance.
+     *
+     * \param num_bits CRC size in bits
+     * \param poly CRC polynomial, in MSB-first notation
+     * \param initial_value Initial register value
+     * \param final_xor Final XOR value
+     * \param input_reflected true if the input is LSB-first, false if not
+     * \param result_reflected true if the output is LSB-first, false if not
+     */
+    crc(unsigned num_bits,
+        uint64_t poly,
+        uint64_t initial_value,
+        uint64_t final_xor,
+        bool input_reflected,
+        bool result_reflected);
+    ~crc();
+
+    /*!
+     * \brief Computes a CRC
+     *
+     * \param data the input data for the CRC calculation
+     * \param len the length in bytes of the data
+     */
+    uint64_t compute(const uint8_t* data, std::size_t len);
```

Comment:
It might be useful to have a stream like interface which doesn’t require *all* of the data to be available, e.g. `push_byte()/push_bytes()` followed by a `finalize()`.

Reply:
I see that it might be useful, and often CRC libraries have this. However, for the intended use case it's not necessary, since all the data is already available at once in a PDU, so I prefer not to add it at the moment. If a new use case for this pops up where not all the data is available at once, then I think it's easy to add the required functions to this class. The `compute()` method would become a wrapper over `push()` followed by `finalize()`.